### PR TITLE
Claim account

### PR DESCRIPTION
**Pull request by aron**
This adds a new `claim` module for allowing users who have registered their usernames to claim their account by providing a password. Included is:

* `h.claim.util.generate_claim_url(request, user_id)` to generate a url for mailout.
* A new view that presents a form where a password can be entered.

The following cases are also covered.

* If the account is unclaimed we present a password form. This updates the user model and logs the user in.
* If the user is logged in, it will redirect to the stream with a message.
* If the user account has already been claimed (it has a password) we redirect to the stream with an error message.

### CODE REVIEW

**Comment by aron** [comment]
This PR needs tests, @nickstenning could you sanity check the implementation in the mean time.

**Comment by nickstenning** [comment]
Looking pretty good.

File: h/claim/util.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+def generate_claim_token(request, user_id):
+    return request.registry.claim_serializer.dumps({'user_id': user_id})
```
Comment (tilgovi) [inline h/claim/util.py]: I think we do 'userid' instead of 'user_id' in most places throughout the code. Let's pick that and stick to it.

File: h/__init__.py [hunk 2]
```diff
@@ -24,7 +29,4 @@ def includeme(config):
     if config.registry.feature('notification'):
         config.include('h.notification')
 
-    config.include('pyramid_jinja2')
-    config.add_jinja2_renderer('.js')
-    config.add_jinja2_renderer('.txt')
-    config.add_jinja2_renderer('.html')
+    config.include('h.claim')
```
Comment (tilgovi) [inline h/__init__.py]: I would add this to the .ini files after h.account rather than include it here. You can just reset this whole file. I like that you've kept this in its own package, but it depends on h.account and h/__init__.py doesn't include that.

**Comment by tilgovi** [comment]
Looks great.

**Comment by tilgovi** [comment]
Small comments.

File: h/claim/test/views_test.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            user = _fake_user()
+            user.password = '<password>'
+            mock_get_by_id.return_value = user
+
+            from ..views import claim_account
+            with raises(HTTPFound):
+                claim_account(request)
+
+    msg = 'This account has already been claimed'
+    mock_flash.assert_called_with(request, msg, kind='error')
```
Comment (aron) [inline h/claim/test/views_test.py]: Could not for the life of me get this patch to work. I've no idea what I'm doing wrong, but the claim_account method runs with the original FlashMessage function.

**Comment by tilgovi** [comment]
It's likely imported directly somewhere, rather than accessed as a module property where it's used. The patch only changes the accessor.

**Comment by aron** [comment]
Aha, so would then it be better to import `horus.lib` and call it via `horuslib.FlashMessage` or something...

**Comment by tilgovi** [comment]
Exactly that would work.

**Comment by nickstenning** [comment]
Rebased, updated, and ready for review.

**Comment by nickstenning** [comment]
@tilgovi, @gergely-ujvari, @csillag: anyone fancy reviewing this?

File: h/claim/views.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    request.registry.notify(LoginEvent(request, user))
+    return exc.HTTPFound(location=request.route_url('stream'))
+
+
+def _validate_request(request):
+    """
+    Check that the passed request is appropriate for proceeding with account
+    claim. Asserts that:
+
+    - no-one is logged in
```
Comment (tilgovi) [inline h/claim/views.py]: Why do we require someone to be logged out to claim another account? Any reason?

**Comment by tilgovi** [comment]
Should we move this whole `claim` package inside the `accounts` package?

File: h/app.py [hunk 5]
```diff
@@ -29,6 +29,9 @@ def create_app(settings):
         api_view = wsgiapp2(api_app)
         config.add_view(api_view, name='api')
 
+    if config.registry.feature('claim'):
```
Comment (tilgovi) [inline h/app.py]: Not sure we need a feature flag. Isn't this just a core feature of accounts?

Comment (tilgovi) [inline h/app.py]: Or it's not because it's only here for us having had a pre-registration period?

Comment (tilgovi) [inline h/app.py]: If that's the case, then I retract my statement about moving this inside the accounts package.

File: h/claim/util.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+def generate_claim_token(request, userid):
+    return request.registry.claim_serializer.dumps({'userid': userid})
```
Comment (tilgovi) [inline h/claim/util.py]: I think I'd prefer to have a function in here, `get_serializer`, that returns the serializer. It's fine that we're just stashing things on the registry object directly rather than using its API to register the serializer, but I think these details should be hidden from other code. `claim.get_serializer(request)` seems better to me.

File: h/templates/layouts/base.html [hunk 7]
```diff
@@ -10,7 +10,13 @@
       {% endfor -%}
     {% endblock %}
 
-    <title>{% block title %}Hypothesis{% endblock %}</title>
+    <title>
+      {%- block title -%}
+        {%- block page_title %}{% endblock -%}
+        {%- if self.page_title() %} | {% endif -%}
```
Comment (tilgovi) [inline h/templates/layouts/base.html]: This is jinja stuff that I don't know about. Can you just clarify for me what's happening here? Is `self` the current template and therefore `self.page_title()` renders that block? Would it be enough to just say `self.page_title` without the function application if we just want to check that the block exists?

File: h/claim/__init__.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+def includeme(config):
+    config.include('.views')
+    config.include('.util')
+
+    # We use the shared session secret, but salt it with the namespace
+    # 'h.claim' -- only messages serialized with this salt will
+    # authenticate on deserialization.
+    secret = config.registry.settings['session.secret']
+    serializer = SignedSerializer(secret, 'h.claim')
+    config.registry.claim_serializer = serializer
```
Comment (tilgovi) [inline h/claim/__init__.py]: The re-use of the session secret, even with a salt, makes me a bit nervous. I'm not an expert in this department, of course. However, the difference between this and the session is that since there is nothing else in the payload other than the user id an attacker could maybe register lots of users for email addresses they own and perhaps execute a chosen plaintext attack, whereas with the session there is the csrf token in there that an attacker cannot select.

Maybe it'd be enough to just stick a random number inside the payload to make it impossible to select the plaintext.

Comment (tilgovi) [inline h/claim/__init__.py]: I fully admit that other mitigations, like rate-limiting sign-up by IP, would be totally effective. I'm just being paranoid and raising the questions, even if I don't have the answers.

Comment (csillag) [inline h/claim/__init__.py]: > I'm just being paranoid and raising the questions, even if I don't have the answers.

Same here. I would like to know a bit more about the background, so that I can be at peace with this.

Comment (tilgovi) [inline h/claim/__init__.py]: http://docs.pylonsproject.org/docs/pyramid/en/latest/narr/security.html#admonishment-against-secret-sharing

Comment (tilgovi) [inline h/claim/__init__.py]: Note that both subsystems use a salt, and yet the docs still advocate not sharing a secret.

Comment (tilgovi) [inline h/claim/__init__.py]: Actually, the auth ticket policy doesn't use a salt. The session does. It's docs say:

    ``salt``
      A namespace to avoid collisions between different uses of a shared
      secret. Reusing a secret for different parts of an application is
      strongly discouraged (see :ref:`admonishment_against_secret_sharing`).
      Default: ``'pyramid.session.'``.

File: h/app.py (hunk 5 above)
Comment (nickstenning) [inline h/app.py]: Yes, it's only for dealing with pre-registration, not the usual account activation processes.

File: h/claim/views.py (hunk 4 above)
Comment (nickstenning) [inline h/claim/views.py]: We automatically log them in at the end of the claim process, so it seems sensible to require that they start from a logged-out state so that they're not confused about who they're logged in as at the end.

File: h/claim/util.py (hunk 6 above)
Comment (nickstenning) [inline h/claim/util.py]: Ok, will do. FWIW I think this was just Aron being [consistent with the way `h.notification` does this at the moment](https://github.com/hypothesis/h/blob/a490e9fea71119304acc54abb77c8b6a28bda9e6/h/notification/__init__.py).

File: h/claim/__init__.py (hunk 8 above)
Comment (nickstenning) [inline h/claim/__init__.py]: Thanks for asking the question. It's good to be paranoid about such things. Just to be clear what this is doing. The "salt" isn't really a salt in any conventional sense. Usually a salt is concatenated to the message in a hashing (or possibly symmetric encryption) operation:

```
    h(M || S)
    E(k, M || S)
```

In this context, the "salt" is actually concatenated to the *key*, in order to form a new key, such that a message `M` serialized with salt `S_1` will not authenticate when passed to the deserializer using salt `S_1`. That is:

```
    HMAC(k || S_1, M) != HMAC(k || S_2, M)
```

So, the question is: how safe is it to derive keys by concatenating a known string to them? In general, the answer is "not very," but for the specific case of HMAC it's probably ok. As far as I'm aware there aren't currently any key recovery or forgery attacks against HMAC-SHA1 anyway.

All of that said, a general cryptographic principle says that keys should be uniform random, so I think I will rework this to derive the keys for these serializers using HMAC instead, so that instead of `k || S`, we use `HMAC(k, S)` as the derived keys for each purpose.

Why not get the deployer to provide multiple keys? Because the faff associated with setting up N keys (the number will only grow over time) actually makes it likely that at some point someone will simply set all the keys to the same thing, which would be *more* dangerous than what we have currently. Or, to put it another way, if we ask deployers to provide *one* strong secret key, then we can derive keys that are *guaranteed* different for different purposes as:

```
    k_session = HMAC(k, S_session)
    k_claim = HMAC(k, S_claim)
    ...
```

Whereas if we ask deployers to provide N different secret keys, we will have no guarantee that the supplied keys are different.

TL;DR: You're probably being overly paranoid, but that's good. I'll rework this to derive keys properly.

File: h/app.py (hunk 5 above)
Comment (nickstenning) [inline h/app.py]: Oh, and IMO all major new chunks of code should be behind a feature flag, so that we can merge self-consistent, well tested, but possibly incomplete work to master, without having to hold up deployments to prod until the feature is complete.

File: h/claim/__init__.py (hunk 8 above)
Comment (tilgovi) [inline h/claim/__init__.py]: I'm not suggesting to remove the h.claim string, just to use a different
secret.

If the operator sets it to the same, your concatenation is the failsafe.

But better for them to set it differently.

Not sure about your derivation idea. Will think after coffee. But I'm
pretty sure just using a different key is the easiest to reason about.

File: h/templates/layouts/base.html (hunk 7 above)
Comment (tilgovi) [inline h/templates/layouts/base.html]: Clarification on this?

**Comment by tilgovi** [comment]
Rebased and made it use the new key derivation. I'll tidy up the registry usage to my preference at some later date. If you could just answer the question about the jinja bit then I think this is good to go.

:shipit:

File: h/templates/layouts/base.html (hunk 7 above)
Comment (aron) [inline h/templates/layouts/base.html]: You're right `self` refers to the template. It's used to call the same block multiple times in a template.

> If you want to print a block multiple times you can however use the special self variable and call the block with that name[1]

We use it here to check whether a child template has provided any content for the `page_title` block and if so render the pipe character to act as a divider. We need to call the function rather than just use `self.page_title` because the block is always defined and returns a string. We just want to know if that string contains anything.

[1] http://jinja.pocoo.org/docs/dev/templates/#child-template

Comment (tilgovi) [inline h/templates/layouts/base.html]: The block is not always defined it seems because it's not an inheritance / overriding of the title block, it's a separate, optional block that the child may define.

I've just tried it. I'm pushing this change.

Comment (tilgovi) [inline h/templates/layouts/base.html]: Or rather, if you simply remove the block definition above this line, then you can test with an `if` whether the child defined it.

Comment (tilgovi) [inline h/templates/layouts/base.html]: 84235988839bd5ced8200ea41b00e90bc364857e
